### PR TITLE
core[patch]:Fix Incorrect listeners parameters for Runnable.with_listeners() and .map()

### LINKED ISSUE #20509: Incorrect listeners parameters for Runnable.with_listeners() and .map()
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

```python
from langchain_core.runnables import RunnableLambda
from langchain_core.tracers.schemas import Run

def fake_chain(inputs: dict) -> dict:
    return {**inputs, "key": "extra"}

def on_start(run: Run):
    print("on_start:", run.inputs)

def on_end(run: Run):
    print("on_end:  ", run.outputs)

chain = RunnableLambda(fake_chain).with_listeners(on_end=on_end, on_start=on_start)
chain = chain.map()

data = [{"name": "one"}, {"name": "two"}]
out = chain.invoke(data, config={"max_concurrency": 1})
print("result:  ", out)
```

`max_concurrency` is added for simplicity.

### Error Message and Stack Trace (if applicable)

_No response_

### Description

I want to store `fake_chain` output using listeners. `with_listeners()` allows to hook only top level runnable (according to its docstring). But `run` object is incorrect if use `map()`.

I expect to see
```
on_start: {'name': 'one'}
on_start: {'name': 'two'}
on_end:   {'name': 'one', 'key': 'extra'}
on_end:   {'name': 'two', 'key': 'extra'}
result:   [{'name': 'one', 'key': 'extra'}, {'name': 'two', 'key': 'extra'}]
```

but get
```
on_start: {'name': 'one'}
on_start: {'name': 'one'}  # <!
on_end:   {'name': 'one', 'key': 'extra'}
on_end:   {'name': 'one', 'key': 'extra'}   # <!
result:   [{'name': 'one', 'key': 'extra'}, {'name': 'two', 'key': 'extra'}]
```

I didn't dive deeper, but smth wrong happens in the `RunnableBindingBase.batch() -> _merge_configs()` (a guess).

### System Info

```shell
$ pip freeze | grep langchain
langchain==0.1.16               
langchain-anthropic==0.1.4      
langchain-community==0.0.33     
langchain-core==0.1.43          
langchain-google-genai==0.0.11  
langchain-google-vertexai==0.1.2
langchain-groq==0.0.1           
langchain-openai==0.0.8         
langchain-text-splitters==0.0.1 
```
platform: `linux`
python: `3.11.8`

### PR DESCRIPTION
**Pull request by liugddx**
- **Issue:** fix #20509
-  @baskaryan, @eyurtsev

image

### CODE REVIEW

**Comment by eyurtsev** [comment]
@liugddx thanks for tackling this! This doesn't look like the correct place for the fix. 


https://github.com/langchain-ai/langchain/blob/0ae5027d9847afaacfbe75d60cf0798e7f46c7e7/libs/core/langchain_core/runnables/base.py#L1433-L1433

Instantiates config factories -- this looks good.

This code is used here:
https://github.com/langchain-ai/langchain/blob/0ae5027d9847afaacfbe75d60cf0798e7f46c7e7/libs/core/langchain_core/runnables/base.py#L4509-L4509

There could be a bug in that place, but most likely the issue is here:

https://github.com/langchain-ai/langchain/blob/0ae5027d9847afaacfbe75d60cf0798e7f46c7e7/libs/core/langchain_core/runnables/base.py#L4262-L4262

We should be likely creating a new config for each member of the batch so that we can generate separate traces 

---

You can take a look at the following python code

```python
def test_listeners() -> None:
    from langchain_core.runnables import RunnableLambda
    from langchain_core.tracers.schemas import Run

    def fake_chain(inputs: dict) -> dict:
        return {**inputs, "key": "extra"}

    def on_start(run: Run):
        print(run)
        print("on_start:", run.inputs)

    def on_end(run: Run):
        print(run) # <-- Yields the same run id, we is not what we want  <- HERE
        print("on_end:  ", run.outputs)

    chain = RunnableLambda(fake_chain).with_listeners(on_end=on_end, on_start=on_start)
    chain = chain.map()

    data = [{"name": "one"}, {"name": "two"}]
    out = chain.invoke(data, config={"max_concurrency": 1})
    print("result:  ", out)

```

**Comment by eyurtsev** [comment]
@liugddx fwiw this is not the easiest fix to make -- when i marked it as help wanted originally I didn't notice that it was part of `RunnableEach`, so it was a simple bug in `batch()` functionality of base Runnable

**Comment by eyurtsev** [comment]
The change that's needed is likely:

```
    def _invoke(
        self,
        inputs: List[Input],
        run_manager: CallbackManagerForChainRun,
        config: RunnableConfig,
        **kwargs: Any,
    ) -> List[Output]:
        configs = [
            patch_config(config, callbacks=run_manager.get_child()) for _ in inputs
        ]
        return self.bound.batch(inputs, configs, **kwargs)
```

In RunnableEach. And the same change for the async variant.

We'd want to verify the fix using unit test:

```

def test_listeners() -> None:
    from langchain_core.runnables import RunnableLambda
    from langchain_core.tracers.schemas import Run

    def fake_chain(inputs: dict) -> dict:
        return {**inputs, "key": "extra"}

    shared_state = {}
    def on_start(run: Run):
        shared_state[run.id] = {
        "inputs": run.inputs
        }


    def on_end(run: Run):
        shared_state[run.id]['outputs'] = run.inputs

    chain = RunnableLambda(fake_chain).with_listeners(on_end=on_end, on_start=on_start)
    chain = chain.map()

    data = [{"name": "one"}, {"name": "two"}]
    out = chain.invoke(data, config={"max_concurrency": 1})
    # FYI: This test might be flaky since listeners if listeners
    # are called in a separate thread
    # We'll probably want to add a very short sleep here to deal with that
    assert len(shared_state) == 2
    # More test code to verify that the outputs are as expected
    
    # Test should also cover the .batch() version 
    
```

**Comment by liugddx** [comment]
> The change that's needed is likely:
> 
> ```python
>     def _invoke(
>         self,
>         inputs: List[Input],
>         run_manager: CallbackManagerForChainRun,
>         config: RunnableConfig,
>         **kwargs: Any,
>     ) -> List[Output]:
>         configs = [
>             patch_config(config, callbacks=run_manager.get_child()) for _ in inputs
>         ]
>         return self.bound.batch(inputs, configs, **kwargs)
> ```
> 
> In RunnableEach. And the same change for the async variant.
> 
> We'd want to verify the fix using unit test:
> 
> ```python
> def test_listeners() -> None:
>     from langchain_core.runnables import RunnableLambda
>     from langchain_core.tracers.schemas import Run
> 
>     def fake_chain(inputs: dict) -> dict:
>         return {**inputs, "key": "extra"}
> 
>     shared_state = {}
>     def on_start(run: Run):
>         shared_state[run.id] = {
>         "inputs": run.inputs
>         }
> 
> 
>     def on_end(run: Run):
>         shared_state[run.id]['outputs'] = run.inputs
> 
>     chain = RunnableLambda(fake_chain).with_listeners(on_end=on_end, on_start=on_start)
>     chain = chain.map()
> 
>     data = [{"name": "one"}, {"name": "two"}]
>     out = chain.invoke(data, config={"max_concurrency": 1})
>     # FYI: This test might be flaky since listeners if listeners
>     # are called in a separate thread
>     # We'll probably want to add a very short sleep here to deal with that
>     assert len(shared_state) == 2
>     # More test code to verify that the outputs are as expected
>     
>     # Test should also cover the .batch() version 
>     
> ```

Thank you for your help

**Comment by liugddx** [comment]
@eyurtsev @baskaryan PTAL

**Comment by liugddx** [comment]
> @eyurtsev @baskaryan PTAL

If there is anything wrong with what I fixed, please let me know, thank you

File: libs/core/tests/unit_tests/runnables/test_runnable.py [hunk 1]
```diff
@@ -5496,3 +5496,67 @@ async def chunk_iterator() -> AsyncIterator[Dict[str, str]]:
...
+        .map()
+    )
+
+    data = [{"name": "one"}, {"name": "two"}]
+    await chain.ainvoke(data, config={"max_concurrency": 1})
+    # FYI: This test might be flaky since listeners if listeners
+    # are called in a separate thread
+    # We'll probably want to add a very short sleep here to deal with that
+    assert len(shared_state) == 2
+    # More test code to verify that the outputs are as expected
```
Comment (eyurtsev) [inline libs/core/tests/unit_tests/runnables/test_runnable.py:5560]: @liugddx did you want to add the additional test code to verify that things are working as expected?

For example, we could verify that the runs have the expected inputs outputs?

Comment (liugddx) [reply libs/core/tests/unit_tests/runnables/test_runnable.py:5560]: > @liugddx did you want to add the additional test code to verify that things are working as expected?
> 
> For example, we could verify that the runs have the expected inputs outputs?

Sorry, I ignored this, I will continue to maintain this PR

**Comment by liugddx** [comment]
Are there any other issues with this PR? please tell me thank you @eyurtsev

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**